pkg/registry/petset/strategy.go

```diff
@@ -0,0 +1,138 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package petset
+
+import (
+	"fmt"
+	"reflect"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/apps"
+	"k8s.io/kubernetes/pkg/apis/apps/validation"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/registry/generic"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/validation/field"
+)
+
+// petSetStrategy implements verification logic for Replication PetSets.
+type petSetStrategy struct {
+	runtime.ObjectTyper
+	api.NameGenerator
+}
+
+// Strategy is the default logic that applies when creating and updating Replication PetSet objects.
+var Strategy = petSetStrategy{api.Scheme, api.SimpleNameGenerator}
+
+// NamespaceScoped returns true because all PetSet' need to be within a namespace.
+func (petSetStrategy) NamespaceScoped() bool {
+	return true
+}
+
+// PrepareForCreate clears the status of an PetSet before creation.
+func (petSetStrategy) PrepareForCreate(obj runtime.Object) {
+	petSet := obj.(*apps.PetSet)
+	// create cannot set status
+	petSet.Status = apps.PetSetStatus{}
+
+	petSet.Generation = 1
+}
+
+// PrepareForUpdate clears fields that are not allowed to be set by end users on update.
+func (petSetStrategy) PrepareForUpdate(obj, old runtime.Object) {
+	newPetSet := obj.(*apps.PetSet)
+	oldPetSet := old.(*apps.PetSet)
+	// Update is not allowed to set status
+	newPetSet.Status = oldPetSet.Status
+
+	// Any changes to the spec increment the generation number, any changes to the
+	// status should reflect the generation number of the corresponding object.
+	// See api.ObjectMeta description for more information on Generation.
+	if !reflect.DeepEqual(oldPetSet.Spec, newPetSet.Spec) {
+		newPetSet.Generation = oldPetSet.Generation + 1
+	}
+
+}
+
+// Validate validates a new PetSet.
+func (petSetStrategy) Validate(ctx api.Context, obj runtime.Object) field.ErrorList {
+	petSet := obj.(*apps.PetSet)
+	return validation.ValidatePetSet(petSet)
+}
+
+// Canonicalize normalizes the object after validation.
+func (petSetStrategy) Canonicalize(obj runtime.Object) {
+}
+
+// AllowCreateOnUpdate is false for PetSet; this means POST is needed to create one.
+func (petSetStrategy) AllowCreateOnUpdate() bool {
+	return false
+}
+
+// ValidateUpdate is the default update validation for an end user.
+func (petSetStrategy) ValidateUpdate(ctx api.Context, obj, old runtime.Object) field.ErrorList {
```

Comment:
Are you asking how technically or how logistically?  Technically, just
return an error

On Tue, Apr 19, 2016 at 5:33 PM, Prashanth B <notifications@github.com>
wrote:

> In pkg/registry/petset/strategy.go
> <https://github.com/kubernetes/kubernetes/pull/24315#discussion_r60335379>
> :
>
> > +func (petSetStrategy) Validate(ctx api.Context, obj runtime.Object) field.ErrorList {
> > +	petSet := obj.(*apps.PetSet)
> > +	return validation.ValidatePetSet(petSet)
> > +}
> > +
> > +// Canonicalize normalizes the object after validation.
> > +func (petSetStrategy) Canonicalize(obj runtime.Object) {
> > +}
> > +
> > +// AllowCreateOnUpdate is false for PetSet; this means POST is needed to create one.
> > +func (petSetStrategy) AllowCreateOnUpdate() bool {
> > +	return false
> > +}
> > +
> > +// ValidateUpdate is the default update validation for an end user.
> > +func (petSetStrategy) ValidateUpdate(ctx api.Context, obj, old runtime.Object) field.ErrorList {
>
> How would we disallow an update to replicas, selectors, labels,
> annotations etc? there's a petsets/status subresource.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24315/files/8ef9b6ad235b4a40e1ac34ef269221a896f4dedd#r60335379>
>